FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+
+    memset( ( void * ) &xCopiedPublish, 0x00, sizeof( xCopiedPublish ) );
+    pxCopiedPublishInfo = &( xCopiedPublish.publishInfo );
+    memcpy( &( xCopiedPublish.publishInfo ), pPublishInfo, sizeof( MQTTPublishInfo_t ) );
+
+    /* Since adding an MQTTPublishInfo_t to a queue will not copy its string buffers,
+     * we need to add buffers to a struct and copy the entire structure. */
+    memcpy( xCopiedPublish.pTopicName, pPublishInfo->pTopicName, pPublishInfo->topicNameLength );
+    memcpy( xCopiedPublish.pPayload, pPublishInfo->pPayload, pPublishInfo->payloadLength );
+    pxCopiedPublishInfo->pTopicName = ( const char * ) xCopiedPublish.pTopicName;
+    pxCopiedPublishInfo->pPayload = xCopiedPublish.pPayload;
+
+    /* Add to response queue. */
+    xQueueSendToBack( pResponseQueue, pxCopiedPublishInfo, DEMO_TICKS_TO_WAIT );
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t prvProcessCommand( Command_t * pxCommand )
+{
+    MQTTStatus_t xStatus = MQTTSuccess;
+    uint16_t usPacketId = MQTT_PACKET_ID_INVALID;
+    bool xAddAckToList = false;
+    MQTTPublishInfo_t * pxPublishInfo;
+    MQTTSubscribeInfo_t * pxSubscribeInfo;
+
+    switch( pxCommand->commandType )
+    {
+        case PROCESSLOOP:
+            LogInfo( ( "Running Process Loop." ) );
+            xStatus = MQTT_ProcessLoop( &globalMqttContext, MQTT_PROCESS_LOOP_TIMEOUT_MS );
+            break;
+
+        case PUBLISH:
+            assert( pxCommand->pContext != NULL );
+            pxPublishInfo = pxCommand->pContext->pPublishInfo;
+            assert( pxPublishInfo != NULL );
+
+            if( pxPublishInfo->qos != MQTTQoS0 )
+            {
+                usPacketId = MQTT_GetPacketId( &globalMqttContext );
+            }
+
+            LogInfo( ( "Publishing message to %.*s.", ( int ) pxPublishInfo->topicNameLength, pxPublishInfo->pTopicName ) );
+            xStatus = MQTT_Publish( &globalMqttContext, pxPublishInfo, usPacketId );
+            pxCommand->pContext->returnStatus = xStatus;
+
+            /* Add to pending ack list, or call callback if QoS 0. */
+            xAddAckToList = ( pxPublishInfo->qos != MQTTQoS0 ) && ( xStatus == MQTTSuccess );
+            break;
+
+        /* TODO: Option to subscribe/unsubscribe without sending a packet,
+         * e.g. for Shadow topics. */
+        case SUBSCRIBE:
+        case UNSUBSCRIBE:
+            assert( pxCommand->pContext != NULL );
+            pxSubscribeInfo = pxCommand->pContext->pSubscribeInfo;
+            assert( pxSubscribeInfo != NULL );
+            assert( pxSubscribeInfo->pTopicFilter != NULL );
+            usPacketId = MQTT_GetPacketId( &globalMqttContext );
+
+            if( pxCommand->commandType == SUBSCRIBE )
+            {
+                LogInfo( ( "Subscribing to %.*s",
+                           pxSubscribeInfo->topicFilterLength,
+                           pxSubscribeInfo->pTopicFilter ) );
```

Comment:
I think it is better to log the topic from the function sending the subscribe or unsubscribe commands. In this demo, it may be just one subscription and this function could be reused to send multiple subscription topics as well

Reply:
I think I'll keep this here and rephrase to indicate this is when the packet is actually sent. Regarding the single subscription, I think I'll edit this demo so that it will only un/subscribe for a single filter at a time, since allowing multiple subscriptions in one packet adds complexity in managing subscriptions beyond what this demo is trying to show

Reply:
Ok, so to clarify, this demo can allow multiple subscriptions per SUBSCRIBE command, or it can check for existing subscriptions before sending SUBSCRIBE packets. I think the latter is more valuable in this demo, since having multiple tasks subscribed to the same topic seems more important to me than being able to send many subscribes in a single packet. It's possible to do both, but it would require some complexity that I think would distract too much from what the demo should be showing. I currently haven't implemented either since I thought it would make more sense in another PR.